modules/images/image-loading-optimization/class-ilo-url-metrics-group-collection.php

```diff
@@ -0,0 +1,241 @@
+		foreach ( $this->breakpoints as $max_width ) {
+			$groups[]  = new ILO_URL_Metrics_Group( array(), $min_width, $max_width, $this->sample_size, $this->freshness_ttl );
+			$min_width = $max_width + 1;
+		}
+		$groups[] = new ILO_URL_Metrics_Group( array(), $min_width, PHP_INT_MAX, $this->sample_size, $this->freshness_ttl );
+		return $groups;
+	}
+
+	/**
+	 * Adds a new URL metric to a group.
+	 *
+	 * Once a group reaches the sample size, the oldest URL metric is pushed out.
+	 *
+	 * @param ILO_URL_Metric $new_url_metric New URL metric.
+	 * @return bool Whether the URL metric was added to a group.
+	 */
+	public function add_url_metric( ILO_URL_Metric $new_url_metric ): bool {
+		foreach ( $this->groups as $group ) {
+			if ( $group->add_url_metric( $new_url_metric ) ) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Gets grouped keyed by the minimum viewport width.
+	 *
+	 * @return ILO_URL_Metrics_Group[] Groups.
+	 */
+	public function get_groups(): array {
+		return $this->groups;
+	}
+
+	/**
+	 * Get group for viewport width.
+	 *
+	 * @throws InvalidArgumentException When there is no group for the provided viewport width. This would only happen if a negative width is provided.
+	 *
+	 * @param int $viewport_width Viewport width.
+	 * @return ILO_URL_Metrics_Group URL metrics group for the viewport width.
+	 */
+	public function get_group_for_viewport_width( int $viewport_width ): ILO_URL_Metrics_Group {
+		foreach ( $this->groups as $group ) {
+			if ( $group->is_viewport_width_in_range( $viewport_width ) ) {
+				return $group;
+			}
+		}
+		throw new InvalidArgumentException(
+			esc_html(
+				sprintf(
+					/* translators: %d is viewport width */
+					__( 'No URL metrics group found for viewport width: %d', 'performance-lab' ),
+					$viewport_width
+				)
+			)
+		);
+	}
+
+	/**
+	 * Checks whether every group is populated with at least one URL metric each.
+	 *
+	 * They aren't necessarily filled to the sample size, however.
+	 * The URL metrics may also be stale (non-fresh). This method
+	 * should be contrasted with the `is_every_group_complete()`
+	 * method below.
+	 *
+	 * @see ILO_URL_Metrics_Group_Collection::is_every_group_complete()
+	 *
+	 * @return bool Whether all groups have some URL metrics.
+	 */
+	public function is_every_group_populated(): bool {
+		foreach ( $this->groups as $group ) {
+			if ( $group->count() === 0 ) {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Checks whether every group is complete.
+	 *
+	 * @see ILO_URL_Metrics_Group::is_complete()
+	 *
+	 * @return bool Whether all groups are complete.
+	 */
+	public function is_every_group_complete(): bool {
+		foreach ( $this->groups as $group ) {
+			if ( ! $group->is_complete() ) {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Gets URL metrics from all groups merged into one list.
+	 *
+	 * @return ILO_URL_Metric[] All URL metrics.
+	 */
+	public function get_merged_url_metrics(): array {
```

Comment:
"merged" to me is a bit of a technicality. Maybe better "flattened", or simply "all"?

Reply:
Went back to flattened in 49ec2c5b.